Full upgrade scenario

iNecas opened the pull request:
This commit intoroduces more structured upgrade scenario consisting of
this scenarios:

  * pre_upgrade_check - non-modifying checks: is the system ready?
  * pre_migrations - modifying procedures before source update
  * migrations - source update and main migrations
  * post_migrations - additional steps after migration
  * post_upgrade_check - non-modifying checks: was the upgrade successful

If there were some failures or warnings in the checks, the reporter
informs the user about it and asks for confirmation.

Things to finish:

  - [x] - `--whitelist` option
  - [x] - full upgrade experience: move some checks to pre/post migrations + add setting of repositories, yum update and insaller run
  - [ ] - add status command
  - [ ] - extract target version and handle situation when some upgrade is already in progress
  - [x] - use `run --phase` instead of `advanced run --phase`

iNecas commented:
Whitelist option added

swapab commented:
Is this expected or the command is incorrect.

```
sat63 :: ~/foreman_maintain ‹upgrade-workflow› » ./bin/foreman-maintain upgrade check --help                                                                                                                   1 ↵
/usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/option/definition.rb:69:in `+': no implicit conversion of Symbol into String (TypeError)
	from /usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/option/definition.rb:69:in `help_lhs'
	from /usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/attribute/definition.rb:27:in `help'
	from /usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/help.rb:82:in `block in add_list'
	from /usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/help.rb:81:in `each'
	from /usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/help.rb:81:in `add_list'
	from /usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/help.rb:49:in `help'
	from /usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/command.rb:81:in `help'
	from /usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/command.rb:139:in `rescue in run'
	from /usr/local/rvm/gems/ruby-2.2.0@foreman_maintain/gems/clamp-0.6.5/lib/clamp/command.rb:131:in `run'
	from ./bin/foreman-maintain:10:in `<main>'
```

kgaikwad commented:
@swapab,  It is not expected. 
Not only for upgrade, it is giving same error for Health command with --help option.

@iNecas ,
I guess it is due to "whitelist" option which is added newly. 
The line `option ['-w', '--whitelist'], :whitelist,`  is causing an issue.

Second argument i.e option argument name should be of String type except if it is not flag.
That's why --assumeyes option it executes without an error as its option arg name is set to `:flag`

iNecas commented:
@swapab @kgaikwad Good catch: updated

iNecas commented:
I've added additional logic to the upgrades, including:
* remembering the state of last upgrade run
* skipping phases if needed (and informing about skipping)
* support for `run_once` steps and `--force` option to enforce running them
* ability to run a specific phase (via `upgrade advanced run --phase`)
* confirmation when getting from pre-upgrade checks to migrations
* more migration proceures, including:
  * turning on maintanence mode (disabling external access to 443 via iptables)
  * setting repositories
  * yum update
  * installer --upgrade

iNecas commented:
@swapab I needed to remove the singleton from ObjectCache, as it was not working well when multiple scenarios were run at once execution, which is what happens in the full upgrade scenario.

iNecas commented:
@swapab please also take a look at https://github.com/theforeman/foreman_maintain/pull/67/commits/074bfd8f2371470815d32bc39288565e59edb32a - I was hitting issue, where the ordering on ruby 1.8.7 was different than on 2.2.2 and it was not respecting the order the steps were added in the scenario.

swapab commented:
@iNecas 
Thanks for refactoring TSort implementation. And good use of `group_by`.

mbacovsky commented:
I've not been able to perform `upgrade list-versions` due to downstream check working only for rhel6.
```diff
diff --git a/lib/foreman_maintain/concerns/system_helpers.rb b/lib/foreman_maintain/concerns/system_helpers.rb
index db21ac0..7d85e51 100644
--- a/lib/foreman_maintain/concerns/system_helpers.rb
+++ b/lib/foreman_maintain/concerns/system_helpers.rb
@@ -59,7 +59,7 @@ module ForemanMaintain
 
       def downstream_installation?
         execute?('rpm -q satellite') ||
-          (execute('rpm -q foreman') =~ /6sat.noarch/)
+          (execute('rpm -q foreman') =~ /[67]sat.noarch/)
       end
 
       def package_version(name)
```

mbacovsky commented on definitions/features/downstream.rb:
```diff
@@ -11,8 +11,39 @@ def current_version
 
+  def rh_repos(sat_version)
+    sat_version = version(sat_version)
+    ["rhel-#{rh_version.major}-server-rpms",
```
My repos on the system are http://pastebin.test.redhat.com/501875

mbacovsky replied:
I was able to continue with using the env vars mentioned in the README.

mbacovsky replied:
Would it make sense to add verification of available repos as a pre-upgrade check to avoid failure during the "changing" phase? Should I file a RFE with this?

mbacovsky commented:
Can missing target version be handled with nicer message? 
```
# ./bin/foreman-maintain upgrade run
/root/foreman_maintain/lib/foreman_maintain/upgrade_runner.rb:47:in `initialize': Unknown version  (RuntimeError)
        from /root/foreman_maintain/lib/foreman_maintain/cli/upgrade_command.rb:16:in `new'
        from /root/foreman_maintain/lib/foreman_maintain/cli/upgrade_command.rb:16:in `upgrade_runner'
        from /root/foreman_maintain/lib/foreman_maintain/cli/upgrade_command.rb:60:in `execute'
        from /usr/share/gems/gems/clamp-0.6.2/lib/clamp/command.rb:67:in `run'
        from /usr/share/gems/gems/clamp-0.6.2/lib/clamp/subcommand/execution.rb:11:in `execute'
        from /usr/share/gems/gems/clamp-0.6.2/lib/clamp/command.rb:67:in `run'
        from /usr/share/gems/gems/clamp-0.6.2/lib/clamp/subcommand/execution.rb:11:in `execute'
        from /usr/share/gems/gems/clamp-0.6.2/lib/clamp/command.rb:67:in `run'
        from /root/foreman_maintain/lib/foreman_maintain/cli.rb:18:in `run'
        from /usr/share/gems/gems/clamp-0.6.2/lib/clamp/command.rb:125:in `run'
        from ./bin/foreman-maintain:10:in `<main>'
```

mbacovsky commented:
Finally I was able to complete 6.1 -> 6.2 upgrade. Everything went fine! :sparkles: 
From UX perspective the output is not very easy to read. With the upgrade it won't fit in default 1k lines terminal buffer so I'd prefer to have the output of the commands just in the log or on the screen only in --verbose mode. Another RFE perhaps.

Besides the few things mentioned above this cool addition is good to go :beers:

iNecas commented on lib/foreman_maintain/cli/upgrade_command.rb:
```diff
@@ -1,59 +1,67 @@
       end
 
-      subcommand 'check', 'Run pre-upgrade checks for upgradeing to specified version' do
+      subcommand 'check', 'Run pre-upgrade checks for upgrading to specified version' do
```
What about `before upgrading`?

iNecas commented on lib/foreman_maintain/concerns/metadata.rb:
```diff
@@ -90,6 +90,12 @@ def for_feature(feature_label)
           end
         end
 
+        # Ensure to not run the step twice: expects the scenario to be presisted
```
It's saved to data.yml, the use-case is, when there is a failure in the middle of the upgrade.

iNecas commented on lib/foreman_maintain/runner.rb:
```diff
@@ -1,48 +1,73 @@
+    # rubocop:disable Metrics/MethodLength, Metrics/AbcSize, Metrics/CyclomaticComplexity
+    def run_scenario(scenario, confirm = true)
+      return if scenario.steps.empty?
+      raise 'The runner is already in quit state' if quit?
```
This is more of a internal developer assertion, that we should not get into, unless there is some bug

iNecas commented on lib/foreman_maintain/upgrade_runner.rb:
```diff
@@ -0,0 +1,189 @@
+        @reporter.before_scenario_starts(scenario)
+        @reporter.puts <<-MESSAGE.strip_heredoc
+          Skipping #{skipped_phase} phase as the last active phase was #{skipped_phase}.
+          To enforce to run the phase, use `upgrade advanced run --phase #{phase}`
```
Changing to `Skipping #{skipped_phase} phase as it was already run before.` It can happen, when the migration fails the first time, and one runs the upgrade later to continue.

iNecas commented on lib/foreman_maintain/cli/base.rb:
```diff
@@ -60,6 +69,15 @@ def self.tags_option
+
+        option ['-w', '--whitelist'], 'whitelist',
+               'Comma-separated list of labels of steps to be ignored' do |whitelist|
+          raise ArgumentError, 'value not specified' if whitelist.nil? || whitelist.empty?
```
Clamp requires to raise ArgumentError here to print nice error to the user, therefore we need to use this here

iNecas commented:
Thanks @mbacovsky for review. I've finally got to address the comments. Before merging, I would like to update the target version from parameter to option (as clamp doesn't handle cases, where you do `upgrade run 6.2 --phase=pre-upgrade-check`, as it expects the parameters at the end, which is  not very intuitive. Also would like to add status command to see what's the current state of the upgrade and making the target version required only when no upgrade is currently in progress.

iNecas commented:
The comment about the verbosity of the output is also valid, but I would like to address it after this PR is merged.

iNecas commented:
I've changed the way we pass the version to use `--target-version` to avoid the issues with interleaving parameters and options in clamp.  Also moved the `run --phase` outside of the `advanced` subcommand and added more information on how the upgrade run works and info about specific phases to the README.

iNecas commented:
Tracking additional features around the upgrade status display http://projects.theforeman.org/issues/20596 and additional checks about currnetly running upgrades http://projects.theforeman.org/issues/20597. Merging this PR to unblock work on the upgrades scenario

swapab commented:
Milestone achieved :checkered_flag:
